lib/protocol/Auth.js

```diff
@@ -150,3 +150,16 @@ Auth.int32Read = function(buffer, offset){
        + (buffer[offset + 2] << 8)
        + (buffer[offset + 3]);
 };
+
+Auth.tokenByPlugin = function(pluginName, pluginData, password) {
+  switch (pluginName) {
+    case 'mysql_native_password':
+      // mysql_native_password only need first 20 bytes scremble.
+      return this.token(password, pluginData.slice(0, 20));
+    case 'mysql_old_password':
+      // mysql_old_password only need first 8 bytes scremble.
+      return this.scramble323(pluginData.slice(0, 8), password);
+    default:
+      throw 'Unknown auth plugin: ' + pluginName;
```

Comment:
This should throw an error object, with a proper code and fatal flag so users can handle it like the other errors.